### PR TITLE
ExponentialMatrixCoefficient class

### PR DESCRIPTION
**Pull request by brendankeith**
This PR introduces the `ExponentialMatrixCoefficient` class as well as `DenseMatrix::Exponential()`.

This will be used in a forthcoming MFEM example for the Monge–Ampère equation.

NOTE: Only 2D is supported at this time. 3D support is planned for a future PR.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4231](https://github.com/mfem/mfem/pull/4231) | @brendankeith | @tzanio | @bslazarov + @psocratis | 4/17/24 | 5/26/24 | 5/28/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: linalg/densemat.cpp [hunk 1]
```diff
@@ -532,6 +532,69 @@ MatrixInverse *DenseMatrix::Inverse() const
...
+      {
+         /// Formulas from Corollary 2.4 of doi:10.1109/9.233156
+         /// Note typo in the paper, in the prefactor in the equation under (i).
+         real_t a = data[0];
+         real_t b = data[1];
+         real_t c = data[2];
+         real_t d = data[3];
+         real_t e = (a - d)*(a - d) + 4*b*c;
+         real_t f = std::exp((a + d)/2.0);
+         real_t g = std::sqrt(std::abs(e)) / 2.0;
```
Comment (dohyun-cse) [inline linalg/densemat.cpp:560]:
```suggestion
         const real_t a = data[0];
         const real_t b = data[1];
         const real_t c = data[2];
         const real_t d = data[3];
         const real_t e = (a - d)*(a - d) + 4*b*c;
         const real_t f = std::exp((a + d)/2.0);
         const real_t g = std::sqrt(std::abs(e)) / 2.0;
```

**Review by dohyun-cse — COMMENTED** [review]
Other than unit test, it looks good to me :)

**Comment by brendankeith** [comment]
After talking to @psocratis, I've decided to add one more new function to this PR. I removed the "ready for review" label for now

File: fem/coefficient.hpp [hunk 2]
```diff
@@ -1334,6 +1334,46 @@ class MatrixArrayCoefficient : public MatrixCoefficient
...
+   using MatrixCoefficient::Eval;
+
+   /// Evaluate coefficient located at the i-th row of the matrix using integration
+   /// point @a ip.
+   void Eval(int i, Vector &V, ElementTransformation &T,
+             const IntegrationPoint &ip);
+
+   /// Evaluate the matrix coefficient @a ip.
+   virtual void Eval(DenseMatrix &K, ElementTransformation &T,
+                     const IntegrationPoint &ip);
```
Comment (dohyun-cse) [inline fem/coefficient.hpp:1372]:
```suggestion
   void Eval(DenseMatrix &K, ElementTransformation &T,
                     const IntegrationPoint &ip) override;
```

File: fem/coefficient.cpp [hunk 3]
```diff
@@ -924,6 +924,78 @@ void MatrixArrayCoefficient::Eval(DenseMatrix &K, ElementTransformation &T,
...
+}
+
+void MatrixArrayVectorCoefficient::Eval(DenseMatrix &K,
+                                        ElementTransformation &T,
+                                        const IntegrationPoint &ip)
+{
+   K.SetSize(height, width);
+   for (int i = 0; i < height; i++)
+   {
+      Vector V(width);
```
Comment (dohyun-cse) [inline fem/coefficient.cpp:990]: Avoid creating vector in a loop

```suggestion
   Vector V(width);
   for (int i = 0; i < height; i++)
   {
```

**Review by dohyun-cse — APPROVED** [review]

**Review by alizma — APPROVED** [review]
Looks good to me!

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
@psocratis and @bslazarov, can you please review?

**Review by psocratis — APPROVED** [review]

**Review by bslazarov — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**